File: engine/sdk/ts-kit-engine/src/sync.ts
export interface RemoteStatus {
  name: string;
  url: string;
  mode: "push" | "pull" | "both";
  filter?: string;
  connected?: boolean;
  last_sync?: string | null;
  pending_diffs?: number;
  last_error?: string | null;
  lag_ms?: number;
}

export interface RemoteConfig {
  name: string;
  url: string;
  mode?: "push" | "pull" | "both";
  filter?: string;
}

export class SyncClient {
  constructor(private baseURL: string, private token?: string) {}

  private headers(): Record<string, string> {
    const headers: Record<string, string> = { "Content-Type": "application/json" };
    if (this.token) headers["Authorization"] = `Bearer ${this.token}`;
    return headers;
  }

  async addRemote(name: string, url: string, mode: "push" | "pull" | "both" = "both", filter = ""): Promise<RemoteConfig> {
    const res = await fetch(`${this.baseURL}/sync/remotes`, {
      method: "POST",
      headers: this.headers(),
      body: JSON.stringify({ name, url, mode, filter }),
    });
    if (!res.ok) throw new Error(`addRemote: ${res.status}`);
    return (await res.json()) as RemoteConfig;
  }

  async removeRemote(name: string): Promise<void> {
    const res = await fetch(`${this.baseURL}/sync/remotes/${encodeURIComponent(name)}`, {
      method: "DELETE",
      headers: this.token ? { Authorization: `Bearer ${this.token}` } : undefined,
    });
    if (!res.ok) throw new Error(`removeRemote: ${res.status}`);
  }

  async status(): Promise<RemoteStatus[]> {
    const res = await fetch(`${this.baseURL}/sync/status`);
    if (!res.ok) throw new Error(`sync status: ${res.status}`);
    const payload = (await res.json()) as { remotes: RemoteStatus[] };
    return payload.remotes;
  }
}

File: engine/sdk/py-kit-engine/kit_engine/sync.py
"""Sync operations client."""

from __future__ import annotations

import requests


class SyncClient:
    """REST client for kit sync endpoints."""

    def __init__(self, base_url: str, token: str | None = None):
        self._url = f"{base_url}/sync"
        self._token = token

    def _headers(self) -> dict[str, str]:
        if not self._token:
            return {}
        return {"Authorization": f"Bearer {self._token}"}

    def add_remote(
        self,
        name: str,
        url: str,
        mode: str = "both",
        filter: str = "",
    ) -> dict:
        resp = requests.post(
            f"{self._url}/remotes",
            json={"name": name, "url": url, "mode": mode, "filter": filter},
            headers=self._headers(),
        )
        resp.raise_for_status()
        return resp.json()

    def remove_remote(self, name: str) -> None:
        resp = requests.delete(f"{self._url}/remotes/{name}", headers=self._headers())
        resp.raise_for_status()

    def status(self) -> dict:
        resp = requests.get(f"{self._url}/status")
        resp.raise_for_status()
        return resp.json()

    def push(self, diffs: list[dict]) -> dict:
        resp = requests.post(f"{self._url}/push", json=diffs, headers=self._headers())
        resp.raise_for_status()
        return resp.json()

    def pull(
        self,
        *,
        since_physical: int,
        since_logical: int,
        since_node: str,
    ) -> list[dict]:
        resp = requests.get(
            f"{self._url}/pull",
            params={
                "since_physical": since_physical,
                "since_logical": since_logical,
                "since_node": since_node,
            },
        )
        resp.raise_for_status()
        return resp.json()
